algorithms/HPdclassifier/R/hpdegbm.R

```diff
@@ -0,0 +1,288 @@
+
+   if(missing(distribution))
+	stop("'distribution' is a required argument")
+
+   if(n.trees <= 0)
+        stop("'n.trees' must be more than 0")
+
+   if(interaction.depth <= 0)
+        stop("'interaction.depth' must be more than 0")
+
+
+   if ((interaction.depth%%1 < 0) && ((interaction.depth%%1 > 0)))
+        stop("'interaction.depth' must be an integer")
+
+   if(n.minobsinnode <= 0)
+        stop("'n.minobsinnode' must be more than 0")
+
+   if ((n.minobsinnode%%1 < 0) && ((n.minobsinnode%%1 > 0)))
+        stop("'n.minobsinnode' must be an integer")
+
+   if(shrinkage <= 0)
+        stop("'shrinkage' must be more than 0")
+
+   if ((shrinkage < 0.001) && ((shrinkage > 1)))
+        stop("'shrinkage' must be between [0.001,1]")
+
+
+   if(bag.fraction <= 0)
+        stop("'bag.fraction' must be more than 0")
+
+   if (bag.fraction > 1)
+        stop("'shrinkage' must be (0,1]")
+
+
+   # if trace=TRUE, print out running time
+   if(trace) {
+        cat("Start model training\n")
+        starttime <- Sys.time()
+    }
+
+   # model training
+   if ((!is.dframe(X_train)) && (!is.darray(X_train))) { # for small data, load the whole data into every core
+      # system parameters are transfered into foreach
+      foreach(i, 1:nExecutor, function(dGBM_modeli=splits(dl_GBM_model,i), best.iter=splits(dbest.iter,i), x=X_train,y=Y_train, 
+                  n.trees=n.trees, distribution=distribution, interaction.depth=interaction.depth, n.minobsinnode=n.minobsinnode, 
+                  shrinkage=shrinkage, bag.fraction=bag.fraction) {
+      library(gbm)
+
+      if (distribution=="multinomial") {
+         y <- unlist(y) #  convert it back to "factor" for multinomial distribution
+      }
+
+      # apply gbm.fit for GBM modeling: local GBM model
+      dGBM_model <- gbm.fit(x, y,  
+         offset = NULL, 
+         misc = NULL, 
+         n.trees = n.trees, 
+         distribution = distribution,
+         w = NULL,
+         var.monotone = NULL,
+         interaction.depth = interaction.depth, 
+         n.minobsinnode = n.minobsinnode,
+         shrinkage = shrinkage, 
+         bag.fraction = bag.fraction, 
+         nTrain = NULL,
+         train.fraction = NULL, 
+         keep.data = TRUE,
+         verbose = TRUE,
+         var.names = NULL,
+         response.name = "y",
+         group = NULL)
+
+       # estimate the best iteration
+       best.iter0 <- gbm.perf(dGBM_model, 
+              plot.it = FALSE, 
+              oobag.curve = FALSE, 
+              overlay = FALSE, 
+              method="OOB")
+
+
+       # gbm.more and gbm.perf have bugs for multinomial distribution 
+       best.iter <- as.matrix(best.iter0)
+
+
+       # convert dGBM_model to a list
+       dGBM_modeli <- list(dGBM_model) 
+  
+       update(dGBM_modeli)
+       update(best.iter)
+      })
+    } else{  ### For big data: load each partition into every core
+     # X_train: dframe/darray
```

Comment:
Will design a new automatic sample size planning logic. If the distributed data is not big, increase the sample rate.